### PR TITLE
Allow serialization of ModelLabelsFeatureswith sparse features

### PR DESCRIPTION
**Pull request by Dekken**

### CODE REVIEW

File: lib/include/tick/array/array.h [hunk 1]
```diff
@@ -526,48 +530,28 @@ CEREAL_LOAD_FUNCTION_NAME(Archive &ar, Array<T> &arr) {
...
+#define ARRAY_DEFINE_TYPE_BASIC(TYPE, NAME)             \
   typedef Array<TYPE> Array##NAME;                      \
   typedef std::vector<Array##NAME> Array##NAME##List1D; \
   typedef std::vector<Array##NAME##List1D> Array##NAME##List2D
 
+#define ARRAY_DEFINE_TYPE(TYPE, NAME)  \
+  ARRAY_DEFINE_TYPE_BASIC(TYPE, NAME); \
+  CEREAL_REGISTER_TYPE(Array##NAME)
+
 ARRAY_DEFINE_TYPE(double, Double);
```
Comment (Mbompr) [inline lib/include/tick/array/array.h]: Can we just comment here that we use ARRAY_DEFINE_TYPE_BASIC for all types that don't need serialization yet?

Comment (Mbompr) [reply lib/include/tick/array/array.h]: And add that's for compile time reasons

File: lib/include/tick/array/sparsearray2d.h [hunk 2]
```diff
@@ -106,6 +98,103 @@ class SparseArray2d : public BaseArray2d<T> {
...
+
+      this->_data = s_data;
+      this->_indices = s_indices;
+      this->_row_indices = s_row_indices;
+
+      this->is_data_allocation_owned = 1;
+      this->is_indices_allocation_owned = 1;
+      this->is_row_indices_allocation_owned = 1;
+    } catch (const std::exception &e) {
+      std::cerr << e.what() << std::endl;
```
Comment (Mbompr) [inline lib/include/tick/array/sparsearray2d.h]: Should we catch this error? Won't this create an "unstable" object?

File: lib/include/tick/array/sparsearray2d.h [hunk 3]
```diff
@@ -106,6 +98,103 @@ class SparseArray2d : public BaseArray2d<T> {
...
+      ar(this->_size);
+
+      inner_save(ar);
+
+      std::vector<INDICE_TYPE> v_indices(this->_indices, this->_indices + this->_size_sparse),
+          v_row_indices(this->_row_indices, this->_row_indices + (this->_n_rows + 1));
+      ar(v_indices);
+      ar(v_row_indices);
+    } catch (const std::exception &e) {
+      std::cerr << e.what() << std::endl;
```
Comment (Mbompr) [inline lib/include/tick/array/sparsearray2d.h]: same here

File: lib/include/tick/array/abstractarray1d2d.h [hunk 4]
```diff
@@ -14,14 +14,32 @@
...
+#include <memory>
 #include <type_traits>
 #include <typeinfo>
 
+#include "alloc.h"
 #include "promote.h"
 #include "vector_operations.h"
 
+// clang-format off
+// Don't touch this!
```
Comment (Mbompr) [inline lib/include/tick/array/abstractarray1d2d.h]: Why is this necessary already? Cereal headers raise warnings? Maybe it worths noting it here

Comment (Dekken) [reply lib/include/tick/array/abstractarray1d2d.h]: It was necessary before I added "SortIncludes: False" to the clang format rules, re-arranging includes is not recommended sometimes

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: we might want to test a bad file tbh - loading is more error prone than saving

File: lib/include/tick/array/array.h (hunk 1 above)
Comment (Dekken) [reply lib/include/tick/array/array.h]: sure,

File: lib/include/tick/array/abstractarray1d2d.h (hunk 4 above)
Comment (Dekken) [reply lib/include/tick/array/abstractarray1d2d.h]: still needed it seems,

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: loading a bad file causes a segfault when the exception is not caught

File: lib/include/tick/array/sparsearray2d.h (hunk 3 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: an exception on saving is unlikely, but sometimes better safe than sorry

perhaps we may want a "Array.isValid()" or ".isLoaded()" that exists as a flag for this

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: perhaps allowing the segfault on load is the best until we decide how to proceed
